src/pydrex/geometry.py

```diff
@@ -0,0 +1,170 @@
+
+    Spherical coordinate convention:
+    - ϕ is the longitude (“azimuth”) ∈ (0, π]
+    - θ is the colatitude (“inclination”) ∈ (0, 2π]
+
+    Returns a tuple containing arrays of ϕ and θ values.
+
+    """
+    x = np.atleast_1d(x).astype(float)
+    y = np.atleast_1d(y).astype(float)
+    z = np.atleast_1d(z).astype(float)
+    r = np.sqrt(x**2 + y**2 + z**2)
+    return (r, np.arccos(z / r), np.sign(y) * np.arccos(x / np.sqrt(x**2 + y**2)))
+
+
+def poles(orientations, ref_axes="xz", hkl=[1, 0, 0]):
+    """Extract 3D vectors of crystallographic directions from orientation matrices.
+
+    Expects `orientations` to be an array with shape (N, 3, 3).
+    The optional arguments `ref_axes` and `hkl` can be used to change
+    the global reference axes and the crystallographic direction respectively.
+    The reference axes should be given as a string of two letters,
+    e.g. "xz" (default), which specify the second and third axes
+    of the global right-handed reference frame. The third letter in the set "xyz"
+    determines the first axis. The `ref_axes` will therefore become the
+    horizontal and vertical axes of pole figures used to plot the directions.
+
+    """
+    upward_axes = next((set("xyz") - set(ref_axes)).__iter__())
+    axes_map = {"x": 0, "y": 1, "z": 2}
+
+    # Get directions in the right-handed frame.
+    directions = np.tensordot(orientations.transpose([0, 2, 1]), hkl, axes=(2, 0))
+    directions_norm = la.norm(directions, axis=1)
+    directions[:, 0] /= directions_norm
+    directions[:, 1] /= directions_norm
+    directions[:, 2] /= directions_norm
+
+    # Rotate into the chosen reference frame.
+    zvals = directions[:, axes_map[upward_axes]]
+    yvals = directions[:, axes_map[ref_axes[1]]]
+    xvals = directions[:, axes_map[ref_axes[0]]]
+    return xvals, yvals, zvals
+
+
+def lambert_equal_area(xvals, yvals, zvals):
+    """Project axial data from a 3D sphere onto a 2D disk.
+
+    Project points from a 3D sphere of radius 1, given in Cartesian coordinates,
+    to points on a 2D disk using a Lambert equal area azimuthal projection.
+    Returns arrays of the X and Y coordinates in the unit disk.
+
+    This implementation first maps all points onto the same hemisphere,
+    and then projects that hemisphere onto the disk.
+
+    """
+    xvals = np.atleast_1d(xvals).astype(float)
+    yvals = np.atleast_1d(yvals).astype(float)
+    zvals = np.atleast_1d(zvals).astype(float)
+    # One reference for the equation is Mardia & Jupp 2009 (Directional Statistics),
+    # where it appears as eq. 9.1.1 in spherical coordinates,
+    #   [sinθcosφ, sinθsinφ, cosθ].
+
+    # First we move all points into the upper hemisphere.
+    # See e.g. page 186 of Snyder 1987 (Map Projections— A Working Manual).
+    # This is done by taking θ' = π - θ where θ is the colatitude (inclination).
+    # Equivalently, in Cartesian coords we just take abs of the z values.
+    zvals = abs(zvals)
```

Comment:
Does the 0 value of `zvals` initially correspond to the North Pole and all other values are negative?

Reply:
z value of +1 is the "north pole" or the upward pointing axis. 0 is at the equator. -1 at the "sout pole", so all negative values are flipped to make the 3d vectors all point to the upper hemisphere, if that makes sense.